app/src/org/commcare/activities/CommCareSetupActivity.java

```diff
@@ -756,7 +756,7 @@ private void launchNextActivityOnAppInstall() {
                 Intent i = new Intent(getApplicationContext(), DispatchActivity.class);
                 startActivity(i);
             }
-        } else {
+        } else if (getCallingActivity().getPackageName().startsWith(BuildConfig.APPLICATION_ID)){
             Intent i = new Intent(getIntent());
             setResult(RESULT_OK, i);
         }

```

Comment:
we should add another else statement and fail there with an error message like - 'Invalid CommCare launch, External applications should use action `android.intent.action.VIEW` to launch CommCare'. You can utilise the existing `fail` method in this class to do that.

Reply:
Noted. I battled a little with that, I was divided between notifying the user and letting it die silently.